### PR TITLE
[GH-490] Channel withdraw and deposit

### PR DESCRIPTION
**Pull request by grzegorz225**
This PR implements the channel withdraw and deposit transaction. Tests for the new code were written. Adjusted ChannelStateOnChain slightly to directly modify the total amount of tokens in the channel.

### CODE REVIEW

**Comment by thepiwo** [comment]
@grzegorz225 this is compatible to 0.16. right?

File: apps/aecore/lib/aecore/channel/channel_state_peer.ex [hunk 1]
```diff
@@ -293,16 +295,17 @@ defmodule Aecore.Channel.ChannelStatePeer do
...
-           responder_pubkey: responder_pubkey
-         },
+         %ChannelStatePeer{} = peer_state,
          fee,
          nonce,
          type,
          spec
        ) do
-    DataTx.init(type, spec, [initiator_pubkey, responder_pubkey], fee, nonce)
+    if type.chainstate_senders? do
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/channel_state_peer.ex]: What does chainstate_senders? mean?

Comment (cytadela8) [reply apps/aecore/lib/aecore/channel/channel_state_peer.ex]: This was introduced by Channel PR, becouse some transactions don't contain senders in encoding.

Comment (grzegorz225) [reply apps/aecore/lib/aecore/channel/channel_state_peer.ex]: Some transactions do not contain sender information. &chainstate_senders/0 checks whether we need to consult the chain in order to obtain sender information. These transactions must have their senders set to an empty list.

**Comment by grzegorz225** [comment]
@thepiwo This should be compatible with epoch 0.16. To be fully sure we need the binary compatibility tests from @cytadela8 (or try to sync with epoch containing these transactions).

**Comment by thepiwo** [comment]
@grzegorz225 @cytadela8 can you try and report back?

File: apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  @spec senders_from_chainstate(ChannelDepositTx.t(), Chainstate.t()) :: list(binary())
+  def senders_from_chainstate(
+        %ChannelDepositTx{channel_id: channel_id, depositing_account: depositing_account},
+        chainstate
+      ) do
+    with %ChannelStateOnChain{
+           initiator_pubkey: initiator_pubkey,
+           responder_pubkey: responder_pubkey
+         } <- ChannelStateTree.get(chainstate.channels, channel_id) do
+      if initiator_pubkey == depositing_account do
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex]: could you restructure this fragment?

File: apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+          depositing_account: depositing_account,
+          amount: amount,
+          state_hash: state_hash,
+          sequence: sequence
+        },
+        _data_tx
+      ) do
+    new_accounts =
+      AccountStateTree.update(accounts, depositing_account, fn account ->
+        Account.apply_transfer!(account, block_height, -amount)
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex]: -amount is written as amount * -1 if you want to be consistent, not a big deal

File: apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+  end
+
+  @spec is_minimum_fee_met?(DataTx.t(), tx_type_state(), non_neg_integer()) :: boolean()
+  def is_minimum_fee_met?(%DataTx{fee: fee}, _chain_state, _block_height) do
+    fee >= GovernanceConstants.minimum_fee()
+  end
+
+  @spec encode_to_list(ChannelDepositTx.t(), DataTx.t()) :: list()
+  def encode_to_list(
+        %ChannelDepositTx{} = tx,
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/tx/channel_deposit_tx.ex]: Could you match the fields needed here?

File: apps/aecore/lib/aecore/channel/tx/channel_withdraw_tx.ex [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+  @doc """
+  One of the senders in ChannelWithdrawTx is not passed with tx, but is supposed to be retrieved from Chainstate. The senders have to be the withdrawing and acknowledging party.
+  """
+  @spec senders_from_chainstate(ChannelWithdrawTx.t(), Chainstate.t()) :: list(binary())
+  def senders_from_chainstate(
+        %ChannelWithdrawTx{channel_id: channel_id, withdrawing_account: withdrawing_account},
+        chainstate
+      ) do
+    with %ChannelStateOnChain{
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/tx/channel_withdraw_tx.ex]: can you restructure this as well?

File: apps/aecore/lib/aecore/channel/tx/channel_withdraw_tx.ex [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+  end
+
+  @spec is_minimum_fee_met?(DataTx.t(), tx_type_state(), non_neg_integer()) :: boolean()
+  def is_minimum_fee_met?(%DataTx{fee: fee}, _chain_state, _block_height) do
+    fee >= GovernanceConstants.minimum_fee()
+  end
+
+  @spec encode_to_list(ChannelWithdrawTx.t(), DataTx.t()) :: list()
+  def encode_to_list(
+        %ChannelWithdrawTx{} = tx,
```
Comment (d-velev) [inline apps/aecore/lib/aecore/channel/tx/channel_withdraw_tx.ex]: same as above, matching the fields here would be nice

File: apps/aecore/lib/aecore/channel/channel_state_peer.ex (hunk 1 above)
Comment (d-velev) [reply apps/aecore/lib/aecore/channel/channel_state_peer.ex]: Ok, thanks

**Comment by DanielaIvanova** [comment]
@grzegorz225 
Could you please provide some examples of how to test channel withdraw and deposit.

**Comment by cytadela8** [comment]
Also documentation should be updated both: usage documentation and compatibility testing documentation.

**Comment by grzegorz225** [comment]
@thepiwo @cytadela8 
Yesterday I verified that the withdraw and deposit transaction are compatible with epoch - epoch is able to gossip and mine these transactions. Today I will run additional tests to check if epoch accepts these transactions as payloads in solo_close_tx and slash_tx.

**Comment by grzegorz225** [comment]
@thepiwo @cytadela8 @d-velev Ready for re-review

**Pull request merged**